runtime/buffer.h

```diff
@@ -0,0 +1,179 @@
+// valid data, it supports a current frame that is guaranteed to be
+// contiguous.  The API comprises routines that establish a given interval
+// of the file as the current frame for reading or writing,
+template<typename STORE> class FileFrame {
+public:
+  using FileOffset = std::uint64_t;
+
+  ~FileFrame() { FreeMemoryAndNullify(buffer_); }
+
+  // The valid data in the buffer begins at buffer_[start_] and proceeds
+  // with possible wrap-around for length_ bytes.  The current frame
+  // is offset by frame_ bytes into that region and is guaranteed to
+  // be contiguous for at least as many bytes as were requested.
+
+  FileOffset FrameAt() const { return fileOffset_ + frame_; }
+  char *Frame() const { return buffer_ + start_ + frame_; }
+  std::size_t FrameLength() const {
+    return std::min(length_ - frame_, size_ - (start_ + frame_));
+  }
+
+  // Returns a short frame at a non-fatal EOF.  Can return a long frame as well.
+  std::size_t ReadFrame(
+      FileOffset at, std::size_t bytes, IoErrorHandler &handler) {
+    Flush(handler);
+    Reallocate(bytes, handler);
+    if (at < fileOffset_ || at > fileOffset_ + length_) {
+      Reset(at);
+    }
+    frame_ = static_cast<std::size_t>(at - fileOffset_);
+    if (start_ + frame_ + bytes > size_) {
+      DiscardLeadingBytes(frame_, handler);
+      if (start_ + bytes > size_) {
+        // Frame would wrap around; shift current data (if any) to force
+        // contiguity.
+        RUNTIME_CHECK(handler, length_ < size_);
+        if (start_ + length_ <= size_) {
+          // [......abcde..] -> [abcde........]
+          std::memmove(buffer_, buffer_ + start_, length_);
+        } else {
+          // [cde........ab] -> [abcde........]
+          auto n{start_ + length_ - size_};  // 3 for cde
+          auto gap{size_ - length_};  // 13 - 5 = 8
+          RUNTIME_CHECK(handler, length_ >= n);
+          std::memmove(buffer_ + n, buffer_ + start_, length_ - n);  // cdeab
+          LeftShiftBufferCircularly(buffer_, length_, n);  // abcde
+        }
+        start_ = 0;
+      }
+    }
+    while (FrameLength() < bytes) {
+      auto next{start_ + length_};
+      RUNTIME_CHECK(handler, next < size_);
+      auto chunk{size_ - next};
+      auto got{
+          Store().Read(fileOffset_ + length_, buffer_ + next, chunk, handler)};
+      length_ += got;
+      RUNTIME_CHECK(handler, length_ < size_);
+      if (got < chunk) {
+        break;  // error or EOF & program can handle it
+      }
+    }
+    return FrameLength();
+  }
+
+  void WriteFrame(FileOffset at, std::size_t bytes, IoErrorHandler &handler) {
+    if (!dirty_ || at < fileOffset_ || at > fileOffset_ + length_ ||
+        start_ + (at - fileOffset_) + bytes > size_) {
+      Flush(handler);
+      fileOffset_ = at;
+      Reallocate(bytes, handler);
+    }
+    dirty_ = true;
+    frame_ = at - fileOffset_;
+    length_ = std::max(length_, frame_ + bytes);
+  }
+
+  bool Flush(IoErrorHandler &handler) {
```

Comment:
This function is declared to return a "bool", but there's no "return" statement.  This seems to be OK with Clang, but not with GNU 8.3.0.

Reply:
Fixed.